feat: enable inter-realm mode in 'gno run'

mvertes opened the pull request:
`gno run` can now run realm packages in inter-realm mode, in standalone local mode. It allows to run inter-realm test programs under the gno debugger.

The realm package path is determined automatically, in order of precedence:

1. the `-pkgpath` flag if set — also useful to simulate running the same code from a different location, like a "package fork",
2. the `// PKGPATH:` directive for (file)test files,
3. the `module` path of the `gnomod.toml` file in the directory,
4. else it falls back to the package name (previous behaviour).

When the resulting package path is a realm path (`gno.land/r/...`), the program is run in inter-realm mode. The `-expr` flag is honored in all cases: the default `main()` works for both `main()` and `main(cur realm)` signatures, and expressions calling a crossing function of the package get the `cur` argument injected automatically (same mechanism as `qeval`).

In the debugger, the current realm is now displayed in line info and stack traces, and is tracked across calls for the `up`, `down` and `stack` commands.

## Application

Executing realm test programs with gno run didn't work:

```console
$ cd gnovm
$ go run ./cmd/gno run ./tests/files/zrealm_std4.gno
panic: crossing() is only allowed in realm packages [recovered]
	panic: std_test/./tests/files/zrealm_std4.gno:9:2: crossing() is only allowed in realm packages:
	--- preprocess stack ---
[...]
```

Now the realm package path is derived from the file's `// PKGPATH:` directive:

```console
$ go run ./cmd/gno run ./tests/files/zrealm_std4.gno
test1 gno.land/r/tests/vm
test2 gno.land/r/tests/vm
test3 gno.land/r/tests/vm
```

No panic and output is identical to test expected output:

```console
$ cd gnovm/pkg/gnolang
$ go test -run TestFiles/zrealm_std4.gno
```

And now we can run the test case under the gno debugger:

```console
$ go run ./cmd/gno run -debug ./tests/files/zrealm_std4.gno
```

Realm packages with a `gnomod.toml` can be run directly too, including local copies of packages already published in the store (package fork):

```console
$ go run ./cmd/gno run -expr 'println(Render(""))' ../examples/gno.land/r/demo/counter
0
```

mvertes committed:
feat: enable inter-realm mode in 'gno run'

A new `-pkgpath` flag is added to `gno run` command in order
to enable inter-realm behaviour in standalone local mode.

It allows to run inter-realm test programs under the gno debugger.

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [x] The pull request description provides enough details (checked by @thehowl)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: mvertes/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🔴 At least one of these user(s) reviewed the pull request: [aronpark1007 davd-gzl jefft0 notJoon omarsy MikaelVallenet] (with state "APPROVED")
    │       ├── 🟢 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)
    └── 🟢 Not (🔴 Pull request author is user: dependabot[bot])

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

mvertes committed:
Display realm at debug status and in call stack.

mvertes committed:
Merge branch 'master' into gno-debug-realm

mvertes committed:
Merge branch 'master' into gno-debug-realm

mvertes committed:
Merge branch 'master' into gno-debug-realm

moul commented on gnovm/cmd/gno/run.go:
```diff
@@ -76,6 +78,13 @@ func (c *runCmd) RegisterFlags(fs *flag.FlagSet) {
+		&c.pkgPath,
+		"pkgpath",
+		"",
+		"value of realm pkgPath, for testing purpose",
```
What if we automatically guess it from the gno.mod file? Alternatively, we could keep this flag but not label it as "for testing." Instead, it could serve as a way to simulate running the same code from a different location, which could make sense, like a "package fork."

mvertes replied:
Yes, if a gno.mod file exists, the `gno run -debug` should be able to derive the default package path to apply. The `-pkgpath` flag is still useful for the simulation use case.

mvertes committed:
Merge branch 'master' into gno-debug-realm

mvertes committed:
Fix debugLookup() by using GetPathForName() and GetPointerTo()

Thank you @jaekwon for your advices, allowing to fix and simplify
drastically this important part of the debugger.

mvertes committed:
Merge branch 'master' into gno-debug-realm

mvertes committed:
Merge branch 'master' into gno-debug-realm

mvertes committed:
debugger: display only position start in locations

mvertes committed:
fix execRun for realms

mvertes committed:
Merge branch 'master' into gno-debug-realm

mvertes committed:
gno run: derive realm pkgpath from gno.mod if present

mvertes committed:
Merge branch 'gno-debug-realm' of github.com:mvertes/gno into gno-debug-realm

mvertes committed:
Merge branch 'master' into gno-debug-realm

mvertes committed:
Merge branch 'master' into gno-debug-realm

mvertes committed:
gno run: set pkgPath from gno.mod only for not main packages

mvertes committed:
fix realm printing in call stack

mvertes committed:
Merge branch 'master' into gno-debug-realm

mvertes committed:
revert parsing gno.mod in gno run

mvertes committed:
Merge branch 'gno-debug-realm' of github.com:mvertes/gno into gno-debug-realm

mvertes committed:
Merge branch 'master' into gno-debug-realm

mvertes committed:
fix previous

mvertes committed:
fix lint

thehowl committed:
Merge remote-tracking branch 'upstream/master' into gno-debug-realm

# Conflicts:
#	gnovm/cmd/gno/run.go
#	gnovm/pkg/gnolang/debugger.go

thehowl committed:
fix(examples/urequire): pass rlm through AbortsContains and PanicsContains

#5673 added these wrappers calling uassert.AbortsContains and
uassert.PanicsContains without the rlm parameter that the uassert
signatures gained on master in the meantime, breaking preprocessing of
urequire/v0 and every package importing it (ci/examples red on master
since 889405c34).

Add the rlm parameter and pass it through, matching the sibling wrappers
in the same file.

thehowl committed:
address review feedback: derive pkgpath, dedup run paths, honor -expr

- Derive the package path automatically from the file's "// PKGPATH:"
  directive (ltzmaxwell) or the gnomod.toml module path (moul, mvertes);
  the -pkgpath flag remains as an explicit override for the "package
  fork" simulation use case.
- Merge the realm and non-realm code paths in execRun (thehowl): realm
  mode now only differs by the transaction store, the origin caller and
  re-setting the active package after realm finalization.
- Honor -expr in realm mode (thehowl): the expression is evaluated with
  MaybeInjectCurForEval, so the default main() works for both crossing
  and non-crossing signatures, and any explicit -expr is honored.
- Construct the run package explicitly instead of resolving it through
  the store pkgGetter, so running local files of a package that also
  exists in the store (e.g. examples/) works as a package fork instead
  of panicking with redeclaration errors.

thehowl committed:
Merge branch 'dev/morgan/fix-urequire-rlm' into gno-debug-realm

thehowl committed:
fix(integration): update stale balance in storage_deposit_price_change

889405c34 (#5673) grew examples/p/nt/urequire/v0, which is part of the
genesis packages deployed by test1 in this test, increasing the genesis
storage deposit by 527,700ugnot (5,277 bytes x 100ugnot/byte) and
pushing the balance below the asserted '9999817' prefix. Master CI
never caught it because the same commit also broke node boot entirely
(urequire preprocessing failure), making the test fail earlier.

The final-balance assertion ('9999854') still holds as the delta does
not cross its prefix boundary.

thehowl commented:
Merging this as it was already reviewed by maxwell, me and moul, and mostly touches tooling.
